### ISSUE TITLE
keeping 

### ISSUE THREAD

**Issue by rgacuity** [Azure/azure-iot-sdk-c#499]
- **OS and version used:** Windows 10

- **SDK version used:** 1.2.3

# Description of the issue:
We're using the "LL" API with AMQP on a Windows 10 device, and can successfully send thousands of messages to our hub, over a few hours.  But we find that after several minutes of inactivity, the connection to the Hub is dropped (see socket failure error below). When using the LL API, does the app on our device need to periodically ping the hub to keep the connection alive?

([This post](https://github.com/Azure/azure-iot-sdk-csharp/issues/46#issuecomment-277864017) from @olivierbloch in the C# SDK forum suggests that that SDK handles those keep-alives automatically)

# Code sample exhibiting the issue:
Our app is based on the iothub_ll_telemetry_sample, but creates and retains a single iothub handle for reuse at later times.

# Console log of the issue:
Error: Time:Wed May 23 07:30:04 2018 File:C:\a\azure-iot-sdk-c\c-utility\adapters\socketio_win32.c Func:socketio_send Line:422 Failure: sending socket failed 10054.
Error: Time:Wed May 23 07:30:04 2018 File:C:\a\azure-iot-sdk-c\c-utility\src\tlsio_schannel.c Func:send_chunk Line:484 xio_send failed
Error: Time:Wed May 23 07:30:04 2018 File:C:\a\azure-iot-sdk-c\c-utility\src\tlsio_schannel.c Func:internal_send Line:525 send_chunk failed
Error: Time:Wed May 23 07:30:04 2018 File:C:\a\azure-iot-sdk-c\c-utility\src\tlsio_schannel.c Func:tlsio_schannel_send Line:1272 send failed
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\saslclientio.c Func:saslclientio_send_async Line:1182 xio_send failed
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\connection.c Func:on_bytes_encoded Line:253 Cannot send encoded bytes
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransport_amqp_common.c Func:on_amqp_connection_state_changed Line:771 amqp_connection was closed unexpectedly; connection retry will be triggered.
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransport_amqp_connection.c Func:on_cbs_error Line:160 CBS Error occured
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransport_amqp_twin_messenger.c Func:on_amqp_messenger_subscription_changed_callback Line:1644 Unexpected AMQP messenger state (rgDevSystem, TWIN_MESSENGER_STATE_STARTED, 0)
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransport_amqp_telemetry_messenger.c Func:process_state_changes Line:1753 messagesender reported unexpected state 0 while messenger was started
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\saslclientio.c Func:saslclientio_send_async Line:1173 send called while not open
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\connection.c Func:on_bytes_encoded Line:253 Cannot send encoded bytes
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\saslclientio.c Func:saslclientio_close_async Line:1122 saslclientio_close called while not open
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\connection.c Func:on_bytes_encoded Line:257 xio_close failed
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransport_amqp_common.c Func:on_amqp_connection_state_changed Line:760 Transport received an ERROR from the amqp_connection (state changed AMQP_CONNECTION_STATE_CLOSED -> AMQP_CONNECTION_STATE_ERROR); it will be flagged for connection retry.
Info: Preparing transport for re-connection
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransportamqp_methods.c Func:iothubtransportamqp_methods_unsubscribe Line:819 unsubscribe called while not subscribed
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\iothub_client\src\iothubtransport_amqp_cbs_auth.c Func:on_cbs_put_token_complete_callback Line:168 CBS reported status code 0, error: '(null)' for put-token operation for device 'rgDevSystem'
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\saslclientio.c Func:saslclientio_send_async Line:1173 send called while not open
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\connection.c Func:on_bytes_encoded Line:253 Cannot send encoded bytes
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\saslclientio.c Func:saslclientio_close_async Line:1122 saslclientio_close called while not open
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\connection.c Func:on_bytes_encoded Line:257 xio_close failed
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\saslclientio.c Func:saslclientio_close_async Line:1122 saslclientio_close called while not open
Error: Time:Wed May 23 07:30:04 2018 File:D:\release_03_02_18\c-sdk-nuget\uamqp\src\connection.c Func:connection_close Line:1358 xio_close failed

**Comment by ewertons** [comment]
Hi @rgacuity ,

yes, the Azure IoT C SDK sends keep-alives to the IoT Hub, however the whole I/O generated and consumed by the SDK depends on calling IoTHubClient_LL_DoWork with a minimum frequency (suggestion at least every 100ms).
If that is not called, the IoT Hub will disconnect the client after 4 minutes of not receiving keep-alives.

Are you calling IoTHubClient_LL_DoWork frequently enough?

**Comment by rgacuity** [comment]
Hi @ewertons,

We've only been calling ...DoWork when we actually had a message we wanted to send.  I guess we'll need to do that more frequently.

**Comment by ewertons** [comment]
Yes, @rgacuity , could you give it a try calling it frequently and see if that addresses the issue you are seeing?

**Comment by rgacuity** [comment]
OK, will do.
